### ISSUE TITLE
The model technically never does 'early exit' ?

### ISSUE THREAD

**Issue by sshkhr** [kalviny/MSDNet-PyTorch#13]
Hi, @kalviny thanks a lot for the excellent implementation! It has been extremely helpful for my work.

Reading through validate() in main.py and dynamic_evaluate() in adaptive_inference.py it seems to me that the model processes the entire input in both anytime prediction and budgeted batch classification setting: 

In main.py in validate() function in Line 227
`output = model(input_var)`

In adaptive_inference.py in calc_logit() function in Line 53:
`output = self.model(input_var)`

which returns an array for results from each block. And then the accuracy calculations for early exit are done in hindsight considering whether the model would have been right or wrong if we consider the output from the intermediate classifiers. Could you please confirm that my interpretation is correct?

**Comment by cpsxhao** [comment]
Yes, this code saves results from each block and "simulates" adaptive inference when calculating the accuracy, rather than performing the "real" early exit process during model inference.

**Comment by binqi-sun** [comment]
Hi @sshkhr and @cpsxhao, thank you for your question and explanation related to the "real early exit" issue. I am also very curious about this, and I am wondering if I can implement the "real early exit" for a dynamic neural network using current DL frameworks like Pytorch and Tensorflow? It would be great if you can give me some clues or show me some example codes for that. Thank you very much!
